testing/sdk_auth.py

```diff
@@ -0,0 +1,376 @@
+    def __init__(self):
+        """
+        Installs the Kerberos Domain Controller (KDC) as the initial step in creating a kerberized cluster.
+        This just passes a dictionary to be rendered as a JSON app defefinition to marathon.
+        """
+        self.temp_working_dir = _create_temp_working_dir()
+        kdc_app_def_path = "{current_file_dir}/../tools/kdc.json".format(
+            current_file_dir=os.path.dirname(os.path.realpath(__file__)))
+        with open(kdc_app_def_path) as f:
+            kdc_app_def = json.load(f)
+
+        kdc_app_def["id"] = KERBEROS_APP_ID
+        _launch_marathon_app(kdc_app_def)
+        _check_kdc_marathon_task_is_running()
+        self.kdc_port = 88
+        self.kdc_fqdn = "{app_id}.marathon.{host_suffix}".format(
+            app_id=KERBEROS_APP_ID, host_suffix=sdk_hosts.VIP_HOST_SUFFIX)
+        self.kdc_host_id = _get_host_id()
+        self.kdc_host_name = _get_host_name(self.kdc_host_id)
+        self.master_public_ip = _get_master_public_ip()
+        self.framework_id = _get_framework_id()
+        self.task_id = _get_task_id()
+        self.principals = []
+        self.keytab_file_name = KERBEROS_KEYTAB_FILE_NAME
+        self.base64_encoded_keytab_file_name = BASE64_ENCODED_KEYTAB_FILE_NAME.format(keytab_name=self.keytab_file_name)
+
+    def add_principals(self, principals: list):
+        """
+        Adds a list of principals to the KDC. A principal is defined as a concatenation of 3 parts
+        in the following order:
+        - primary: first part of the principal. In the case of a user, it's the same as your username.
+                   For a host, the primary is the word host.
+        - instance: The instance is a string that qualifies the primary.
+                    In the case of a user, the instance is usually null, but a user might also have an additional
+                    principal, with an instance called admin.  In the case of a host, the instance is the fully
+                    qualified hostname, e.g., daffodil.mit.edu.
+        - realm: your Kerberos realm. In most cases, your Kerberos realm is your domain name, in upper-case letters.
+
+        More info on principal definition:
+        https://web.mit.edu/kerberos/krb5-1.5/krb5-1.5.4/doc/krb5-user/What-is-a-Kerberos-Principal_003f.html
+
+        A principal is formatted as: <primary>/instance@realm
+        Eg. hdfs/name-0-node.hdfs.autoip.dcos.thisdcos.directory@LOCAL
+
+        :param principals: The list of principals to be added to KDC.
+        """
+        # TODO: Perform sanitation check against validity of format for all given principals and raise an
+        # exception when the format of a principal is invalid.
+        self.principals = principals
+
+        log.info("Adding the following list of principals to the KDC: {principals}".format(principals=principals))
+        add_principal_cmd = "task exec {task_id} /usr/sbin/kadmin -l add --use-defaults --random-password {principal}"
```

Comment:
Later: We could have a `run_kadmin` method to wrap this.

Reply:
Good suggestion, semantically cleaner.